API.md

```diff
@@ -0,0 +1,49 @@
+- GET /documents/:id/edit - displays a form to edit an existing document
+- POST /documents - creates a new document
+- PATCH/PUT /documents/:id - updates an existing document
+- DELETE /documents/:id - deletes an existing document
+
+And a CODAP specific API that was also used for the initial version of the [Cloud File Manager](https://github.com/concord-consortium/cloud-file-manager) that is called via AJAX requests:
+
+- GET /document/all - returns a json array of documents assigned to the owner with the main codap document attribute set.  The format of the returned data is:  ```[{"name": <name>, "id": <id>, _permissions: (1|0)}, ...]``` where ```_permissions: 1``` means the document is shared.
+
+- GET /document/open - returns the content of the document specified by a varying set of query parameters.  The first set is a combination of the recordname or doc parameter that is used as a title lookup combined with the owner's user id.  Those two attributes are then used to lookup the document first with the runKey parameter and then without it if no document is found.  If no recordname or doc parameter is present the recordid parameter is used.  If the user is authorized to open the document but is not the owner then a second document with the same title and runKey but with the current user as the owner is either found or created.  The format of the returned data is the contents of the file with three extra HTTP headers: ```Document-Id, X-Codap-Opened-From-Shared-Document and X-Codap-Will-Overwrite```
+
+- GET /document/save - saves the document specified either by recordid or a combination of title and runKey.  If the document is found but the user is not the owner and is authorized to save then a new document is created using the same title and runKey if it has not already been created.  Along with the document content the shared and parent_id attributes are set on the document if present in the query parameters.
+
+- GET /document/patch - updates a document specified by recordid using the JSON Patch specification.  Along with patching the contents the shared and parent_id attributes are also set if present in the query parameters.
+
+- GET /document/rename - renames a document specified in the same manner as the open request.
+
+- GET /document/delete - deletes a document specified in the same manner as the open request.
+
+- GET /document/launch - renders a launch button for a document specified in the same manner as the open request.
+
+- GET /document/report - renders a launch button for the report view for a document specified in the same manner as the open request.
+
+## Version 2
+
+The second version of the API is handled by the documents_v2 controller and exposes only a non-GUI programmatic API in the /v2/ namespace as:
+
+- GET /v2/documents/:id - returns the contents of a document specified by id.  This requires either the document's read-only or read-write access keys.
```

Comment:
If the document is shared, then neither access key is required.  Also can you add examples for the 3 options here: shared, read-only, read-write.  And the same for the other 3 endpoints.